### PR TITLE
Bucket management support for alibaba cloud

### LINKED ISSUE #1249: Bucket Management support for Alibaba cloud
**Issue/Feature Description:**
Need to implement bucket operations for alibaba cloud.
This should inlcude:
1. Create cloud bucket
2. Delete cloud bucket
3. Upload object to bucket
4. Upload larger object to bucket
5. Create folder in bucket
6. Upload object in folder
7. Delete object from folder
8. Delete folder
9. Delete object
10. Migration to/from support
11. Lifecycle transition to/from

**Why this issue to fixed / feature is needed(give scenarios or use cases):**

**How to reproduce, in case of a bug:**

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

### PR DESCRIPTION
**Pull request by PravinRanjan10**
This PR has implementation of bucket management for alibaba cloud. The code is tested with all bucket operation like, create, delete, upload, multi-part upload, migration etc.

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind new feature
> /kind bug fix
> /kind cleanup
> /kind revert change
> /kind design
> /kind documentation
> /kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1249  

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:
1. Register  backend
ali-backend

2. Create bucket
Create-ali-bucket

3. Upload Object
ali-upload

4. Upload multipart(size 24Mb)
ali-multi-part-upload

5. Migration(aws-->alibaba)
ali-migration

6. Create Folder
ali-Create-Folder

7. Object upload in folder
upload-object-infolder

8. Delete object
ali-delete-object

9. Delete bucket
ali-delete-buck

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by kumarashit** [comment]
Please open corresponding issue and put the details over there. Then attach that to this PR

**Comment by PravinRanjan10** [comment]
> Please open corresponding issue and put the details over there. Then attach that to this PR

Done. Open issues #1249

**Review by kumarashit — APPROVED** [review]
LGTM

**Comment by PravinRanjan10** [comment]
Opend #1254 for UT and coverage improvment

File: s3/pkg/datastore/alibaba/alibaba.go [hunk 1]
```diff
@@ -42,18 +42,34 @@ type OSSAdapter struct {
 }
 
 func (ad *OSSAdapter) BucketDelete(ctx context.Context, in *pb.Bucket) error {
+	log.Info("Bucket delete is called in s3 alibaba-oss")
+	err := ad.client.DeleteBucket(in.Name)
+	if err != nil {
+		log.Error("the bucket deletion failed in s3 alibaba-oss with err:%s", err.Error())
```
[the remark this answers is not in the document]
Comment (rajat-soda) [reply s3/pkg/datastore/alibaba/alibaba.go:48]: check these in rest of the PR

File: s3/pkg/datastore/alibaba/alibaba.go [hunk 2]
```diff
@@ -42,18 +42,34 @@ type OSSAdapter struct {
...
+	log.Debug("The bucket deletion successful in s3 alibaba-oss")
+
 	return nil
 }
 
 func (ad *OSSAdapter) BucketCreate(ctx context.Context, input *pb.Bucket) error {
+	log.Info("Bucket create is called in s3 alibaba-oss")
+	err := ad.client.CreateBucket(input.Name)
+	if err != nil {
+		log.Error("the create bucket failed in s3 alibaba-oss with err:%s", err.Error())
```
Comment (rajat-soda) [inline s3/pkg/datastore/alibaba/alibaba.go:60]: Here too.

File: s3/pkg/datastore/alibaba/alibaba.go [hunk 3]
```diff
@@ -314,8 +330,8 @@ func (ad *OSSAdapter) CompleteMultipartUpload(ctx context.Context, multipartUplo
 
 func (ad *OSSAdapter) AbortMultipartUpload(ctx context.Context, multipartUpload *pb.MultipartUpload) error {
 
-	bucket := ad.backend.BucketName
-	newObjectKey := multipartUpload.Bucket + "/" + multipartUpload.Key
+	bucket := multipartUpload.Bucket
```
Comment (rajat-soda) [inline s3/pkg/datastore/alibaba/alibaba.go:333]: How is the failure of object upload handled

**Review by rajat-soda — CHANGES_REQUESTED** [review]

File: s3/pkg/datastore/alibaba/alibaba.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply s3/pkg/datastore/alibaba/alibaba.go]: err==>error, alibaba==>Alibaba   done!!

in log.error(""), first letter is in lowercase only

File: s3/pkg/datastore/alibaba/alibaba.go (hunk 3 above)
Comment (PravinRanjan10) [reply s3/pkg/datastore/alibaba/alibaba.go:333]: It's just object parameter. so error handling is not required.

**Comment by rajat-soda** [comment]
LGTM

**Review by skdwriting — APPROVED** [review]
LGTM

**Pull request merged**